crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -5939,23 +6095,27 @@ fn perform_rich_comparison<'db>(
 ) -> Result<Type<'db>, CompareUnsupportedError<'db>> {
     // The following resource has details about the rich comparison algorithm:
     // https://snarky.ca/unravelling-rich-comparison-operators/
-    //
-    // TODO: this currently gives the return type even if the arg types are invalid
-    // (e.g. int.__lt__ with string instance should be errored, currently bool)
-
-    let call_dunder = |op: RichCompareOperator,
-                       left: InstanceType<'db>,
-                       right: InstanceType<'db>| {
-        match left.class.class_member(db, op.dunder()) {
-            Symbol::Type(class_member_dunder, Boundness::Bound) => class_member_dunder
-                .call(
-                    db,
-                    &CallArguments::positional([Type::Instance(left), Type::Instance(right)]),
-                )
-                .return_type(db),
-            _ => None,
-        }
-    };
+    // TODO: This should use `call_dunder` instead of reimplementing it from scratch.
+    let call_dunder =
```

Comment:
We can't use `call_dunder` here because `call_dunder` contains a `to_meta_class` call. Maybe we should consider removing it and let the call site deal with converting to a meta class first.

Reply:
I don't think we should do that, because "looked up on the class, not the instance" is part of the contract of calling a dunder method in general in Python. And it is true here as well: see the use of `left.class.class_member` below, which is equivalent (for an Instance type) on looking up a member on the meta type.